creusot/src/ctx.rs

```diff
@@ -22,26 +23,44 @@ use rustc_errors::{DiagnosticBuilder, DiagnosticId};
 use rustc_hir::def_id::{DefId, LocalDefId};
 use rustc_infer::traits::{Obligation, ObligationCause};
 use rustc_middle::{
+    mir::{Body, MirPass, Promoted},
     thir,
     ty::{
         subst::{GenericArgKind, InternalSubsts},
         GenericArg, ParamEnv, SubstsRef, Ty, TyCtxt, WithOptConstParam,
     },
 };
+use rustc_mir_transform::{cleanup_post_borrowck::CleanupPostBorrowck, simplify::SimplifyCfg};
 use rustc_span::{Span, Symbol, DUMMY_SP};
 use rustc_trait_selection::traits::SelectionContext;
 pub(crate) use util::{module_name, ItemType};
 use why3::exp::Exp;
 
 pub(crate) use crate::translated_item::*;
 
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub struct BodyId {
+    def_id: LocalDefId,
+    pub promoted: Option<Promoted>,
+}
+
```

Comment:
FINALLY, GOOD, THANKS. Incidentally, I've seen discussions about giving promoteds their own `DefId`s but this is good.